### ISSUE TITLE
GIL information

### ISSUE THREAD

**Issue by maartenbreddels** [gaogaotiantian/viztracer#60]
Hi,

I was wondering if you thought about visualizing when a thread has or tries to get the GIL?

cheers,

Maarten

**Comment by gaogaotiantian** [comment]
It would be a really nice feature and I did look into it before. However, there's no way(that I know of) to hook the switch of GIL. Also, the front-end(chrome trace viewer) did not support this kind of visualization. Hacking into the front-end for this kind of feature is probably too much at this point. 

One thought that I have for now, is to use a function entry as a suggestion of "holding the GIL", and visualize which thread has it in an optional separate event signal(maybe under `--log_gil`?) that is calculated rather than recorded when the trace is parsed. I could try a prototype this weekend and I'll let you know if it's promising.

**Comment by maartenbreddels** [comment]
What about visualizing it similar to a single thread/process, where each gil lock would be visualized as a separate function call. So if you have pure Python, it would be a single call, and every time the GIL is released, it is 'idle'. The function name could be the thread-id that has the GIL.



> However, there's no way(that I know of) to hook the switch of GIL.

Can you elaborate a bit? Are you saying, you only get called by the profiler machinery, if the thread is not running, you cannot do anything from that thread?

Would it be an idea to do polling? You could have 1 extra thread continuously checking (every X msec, or X usec?) if the GIL is locked, and which thread has it. Looking at https://github.com/python/cpython/blob/5977a7989d49c3e095c7659a58267d87a17b12b1/Python/pystate.c#L1433 and https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Include/internal/pycore_runtime.h#L35 it sounds like it should be possible to figure that out. I guess even doing that ever usec shouldn't be too resource consuming.

You might also be interested in:
https://www.youtube.com/watch?v=HtbLNgXmLrw&ab_channel=EuroPythonConference

**Comment by gaogaotiantian** [comment]
Visualization would be less of a problem. Like I said, having a separate event signal for GIL is doable. The problem is getting data. 

Polling is not a good idea in general, for VizTracer, because it's against VizTracer's fundamental idea. A very important basic of VizTracer is that it's not only a profiler. It is an execution recorder. Polling for GIL might be helpful for a sampling thread profiler, but it will generate confusing data for VizTracer. For example, You can't tell when or if GIL switched between two samples, so it's highly possible that you will show in the report that some thread has the GIL, but the other thread is calling a function(which is not supposed to happen in Python). 

The more frequently you poll, the more overhead you will introduce. Making it meaningful for VizTracer(instead of a sampling profiler) will definitely bring the overhead to a comparable level as the current overhead itself. Basically, it's still a conflict between the idea of polling and VizTracer.

What I can try, as I mentioned earlier, is try to guess GIL with function calls, which is already hooked anyway. Every time a function is called or returned, the thread has the GIL. This way, we will lose nothing while the program is running, and some time when we parse the log(and it would be optional).

**Comment by maartenbreddels** [comment]
I think we wouldn't know in between when the GIL is release (all threads
executing), is that correct?
Maybe the polling could be extra information apart from checking at
function calls. Would you think the overhead is large, if it is polled
every 10usec in c?
I'm surprised there is no API/callback for this.

(from mobile phone)

**Comment by gaogaotiantian** [comment]
I checked the C/Python API again. I'm afraid there's no easy way to even check which thread is hold the GIL. Maybe if I can do some really hacky thing for the data structure, but that may cause a racing condition that I'm not aware of. 10us polling is actually a pretty high frequency. Especially considering that there's thread context switch involved. Also I doubt if you can guarantee a 10us polling because the scheduler might not be able to do it. As a comparison, python GIL switching is 5ms by default. 

I'm leaning towards to giving polling up for the uncertainties:

* The public C/Python API does not give you a way to access current thread that's holding GIL
* Racing condition to access the data is unknown
* Overhead might be high, considering the thread context switch and the polling frequency
* Even if everything works, we still need a thread-safe(no lock) way to store the data to make it meaningful. Sync issues with the circular buffer

Unless someone could come up with a prototype of polling that is

* Safe, won't break the program in any cases
* Reasonably fast, does not increase the overhead by 20%+
* Meaningful
* Working on 3.6 - 3.9 with all OS supported.
 to convince me, I won't move forward on this.

However, I will check the possibility of using the existing data to infer the GIL status. If that's promising I will add it, if not then we'll probably need to wait for someone that's better than me to create the feature.

**Comment by maartenbreddels** [comment]
FYI, py-spy does detect gil, but I just reported it not working
https://github.com/benfred/py-spy/issues/329 🙂

(from mobile phone)

**Comment by gaogaotiantian** [comment]
Yeah I'm aware that py-spy claim that they can do GIL detect. They are doing it in a pretty hacky way. I don't think you are supposed to read _PyRuntime in python extensions and I'm not sure it there will be racing conditions that will bring down the program while you are dereferencing the pointers. I'm kind of opposed to it as VizTracer is running together with the python program.

**Comment by maartenbreddels** [comment]
FYI, I'm looking into using the `perf` to get notified on thread state changes, and with the callstacks I can see if is due to GIL 'take' or 'drop'
```
python 302631 [018] 3485124.202569:       sched:sched_switch: prev_comm=python prev_pid=302631 prev_prio=120 prev_state=S ==> next_comm=swapper/18 next_pid=0 next_prio=120
        ffffffff96ed4785 __sched_text_start+0x375 ([kernel.kallsyms])
        ffffffff96ed4785 __sched_text_start+0x375 ([kernel.kallsyms])
        ffffffff96ed4b92 schedule+0x42 ([kernel.kallsyms])
        ffffffff9654a51b futex_wait_queue_me+0xbb ([kernel.kallsyms])
        ffffffff9654ac85 futex_wait+0x105 ([kernel.kallsyms])
        ffffffff9654daff do_futex+0x10f ([kernel.kallsyms])
        ffffffff9654dfef __x64_sys_futex+0x13f ([kernel.kallsyms])
        ffffffff964044c7 do_syscall_64+0x57 ([kernel.kallsyms])
        ffffffff9700008c entry_SYSCALL_64_after_hwframe+0x44 ([kernel.kallsyms])
            7f2b30a97376 futex_wait_cancelable+0x216 (inlined)
            7f2b30a97376 __pthread_cond_wait_common+0x216 (inlined)
            7f2b30a97376 __pthread_cond_wait+0x216 (inlined)
            55f77685fe17 drop_gil+0x51 (inlined)
            55f7769c2916 _PyEval_EvalFrameDefault+0x6226 (/home/maartenbreddels/miniconda/envs/dev/bin/python3.7)
            55f776931b67 PyEval_EvalFrameEx+0x117 (inlined)
            55f776931b67 function_code_fastcall+0x117 (inlined)
            55f776931b67 _PyFunction_FastCallDict+0x117 (/home/maartenbreddels/miniconda/envs/dev/bin/python3.7)
```

I think that's more promising than trying to do it from userspace.

**Comment by gaogaotiantian** [comment]
That will involve a third party program, which I will try my best to avoid. Also this is linux only. Also I'm not sure about the overhead of perf. 

However, it would be a interesting tool to use something like this to get a precise GIL performance.

**Comment by maartenbreddels** [comment]
Yeah, totally understand that, would be great if viztracer could do without. However, I may want to add in more information, for instance from the kernel (kernel tracing/events) which I want to combine. I think it can be complementary.
A really ugly hack so far produced this: 
![image](https://user-images.githubusercontent.com/1765949/101996116-c53df180-3ccf-11eb-8140-038288f023d3.png)

Each 'R' block is when a thread/process is in the running state. I don't know yet if that is the GIL or not, but already interesting. Visualization can be done better, and it required manual clock synchronization (and seems off)


This is the code that was run:
```python
import threading
import viztracer
import time
import os
import sys

print(os.getpid())


def run():
    total = 0
    for i in range(1_000_000):
        total += i
    return total


with viztracer.VizTracer(output_file="threads-3.json"):
    thread1 = threading.Thread(target=run)
    thread2 = threading.Thread(target=run)
    thread1.start()
    thread2.start()
    time.sleep(0.2)
    for thread in [thread1, thread2]:
        thread.join()
```

**Comment by gaogaotiantian** [comment]
Yeah, one of the problem of using a third party program is actually the clock synchronization. 

The information itself definitely looks interesting. The visualization won't be a problem as long as you are more familiar with the trace event format.(You need the same pid, a unique tid(like 1) for all the kernel data, and different names for each thread). 

If you can somehow fix the synchronization issue(I don't have a clue of how to do that yet) then it seems like a pretty interesting plugin for VizTracer.

BTW, VizTracer already supports combination of reports by `--combine`. If you can generate a standalone json report of kernel usage, you can simply combine with the existing VizTracer json report.

**Comment by maartenbreddels** [comment]
> Yeah, one of the problem of using a third party program is actually the clock synchronization.

Ok, I dove into this a bit, (https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.rgcugfne66rf and https://docs.google.com/document/d/1RPg7O8lHhYFYQA41OWKEYvD81PV6D1KWFv062k0cWhw ) on android they seem to use a version of ftrace to do some synchronization. It's not clear exactly how it works, but my idea was that if `snaptrace_start` would do a call to `sleep`, I could find that moment back in the call stack, and do some synchronization.

However, it's much simpler when combining with `perf`, since it supports settings the clock `$ perf -k CLOCK_MONOTONIC ...` which is the same clock that viztracer uses! problem solved.

I like the idea of this `--combine`, because I can simply do a `perf2traceevent` script, and it would be independent from viztracer. 

It would be nice to include the FlowEvents as well, to signal which process wakes up who (for instance using mutex/semaphore).

Simple counters, such as page faults, memory reads, and cycle and instruction counters (as indicators of mem bandwidth saturation) can be overlayed (here I'm using S for sleeping state).

![image](https://user-images.githubusercontent.com/1765949/102022676-a5b6cf80-3d88-11eb-8832-b744345f7420.png)

**Comment by gaogaotiantian** [comment]
Hmm, now it looks super interesting. if ```perf``` can provide so much information, then it would be an interesting discussion to include it in VizTracer(still working in parallel and automatically combine the reports together). The user can specify like ```--log_perf``` or something like that, then VizTracer can run ```perf``` along with the script, and generate a combined html report. I have considered flow events, there are some minor issues there but I won't say it's impossible. 

I think rich features like this can let us forget that it's Linux only for now.

**Comment by gaogaotiantian** [comment]
Hmm, I realized `perf` does not support WSL2 well. I guess it would interesting to make a third party tool to generate chrome trace event compatible json files and use ```--combine``` to build the report.

**Comment by gaogaotiantian** [comment]
I'm thinking about adding some standard plugin format so people can develop their own tools that generates chrome trace format compatible data, and VizTracer can load the plugin easily and combine the report automatically.

Something like `viztracer --plugin linux-perf "args"`. I probably need to take a look at other plugins to figure out the design and the way to use it.

**Comment by maartenbreddels** [comment]
> Hmm, I realized `perf` does not support WSL2 well.

Not well, or not at all? I saw the hardware counters are not all working, but tracepoints might work (to get the process state).

I like the idea of having a plugin option, but in general i cannot use viztracer from the command line (all the imports add too much to the json). 

Have you thought instead of injecting the profiler in a running progress, sth like
```
$ viztracer record -P <pid> -- sleep 3
```

I know this can be done with gdb.

**Comment by gaogaotiantian** [comment]
You basically need to compile from source to make perf work partially on WSL2. 

If you use ```--ignore_non_file```, you won't record anything from import. Hope that'll help. 

I could design a plug-in architecture for both command line and inline usage. That takes some time. It would be really helpful for you to describe the process you need to get data from perf(and convert), so I can design the plug-in architecture properly to make this happen(and with reasonably good separations). 

I actually thought about remote control recently. Using gdb to inject code is probably risky(I heard that could cause dead-lock in some cases). I think I have some prototype to install the viztracer first(withouth enabling it) and enable it through signals. That's not perfect, because you still can't profile any process you want, but that's something.

**Comment by maartenbreddels** [comment]
Take a look at entrypoints or pluggy for a plugin system, and yes, signals are a way out, but maybe it's also good to have an unsafe option (as a last resort).

I've put the project at https://github.com/maartenbreddels/per4m as a demo, I'm not sure I have the bandwidth to pursue this seriously, but at least it's more reproducable this way.

I'm surprised this GIL visualization problem got so little attention.

**Comment by maartenbreddels** [comment]
Btw, windows has xperf, and osx has instruments, at least it should be possible to get process state out of them, and for sure also hardware performance counters.

**Comment by gaogaotiantian** [comment]
I'll take a look at your implementation when I have some time and figure out if this should be something that I pursue(and how). As for gdb code injection, that probably requires much more work. I won't do it immediately, but I may take a look at it in the future. And as always, anyone that thinks this as an interesting project is welcome to make a prototype for proof of concept, and eventually a PR for it.

**Comment by maartenbreddels** [comment]
As an example
https://github.com/lmacken/pyrasite/blob/d0c90ab38a8986527c9c1f24e222323494ab17a2/pyrasite/injector.py#L22 (warning, it is GPL)

**Comment by maartenbreddels** [comment]
I've made it easy to combine perf and viztracer:
```
$ pip install per4m
 # follow https://github.com/maartenbreddels/per4m#linux-side
$ per4m giltracer -m per4m.example2
```

Which should give sth like this:
![image](https://user-images.githubusercontent.com/1765949/102225240-458a6f80-3ee7-11eb-9130-27bdfc87459d.png)

Thanks for your help on this, this is very useful (feel free to close, if you consider it closed)

**Comment by gaogaotiantian** [comment]
The info you got from perf is interesting. I'll spend some time on it to see if I can make it an internal plugin for viztracer

**Comment by gaogaotiantian** [comment]
Hi, @maartenbreddels , I have added the plugin structure to viztracer in master. I was wondering if you would like to try it out and adapt your current code to a plugin of VizTracer. 

You can install viztracer by ```pip install git+https://github.com/gaogaotiantian/viztracer``` as it's not officially released yet. I tried to describe how to develop a plugin in https://viztracer.readthedocs.io/en/latest/viz_plugin.html and how to use them in https://viztracer.readthedocs.io/en/latest/plugins.html. It's pretty straightforward to me. Also it's not hard to develop one as it's basically a message handler. 

However, I know you must be busy with your own project, so I totally understand that you've already had what you need and won't spend more time on this matter. 

Either way, I'm planning to make some official plugins in the next week or so, for some extra optional features for VizTracer. 

Cheers.

**Comment by gaogaotiantian** [comment]
The plugin feature is released. I have no current plan to implement GIL detection because it requires too much pre-setup. Close it for now.
